fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/RecordConstructorValue.java

```diff
@@ -0,0 +1,246 @@
+ * {@link com.apple.foundationdb.record.query.plan.temp.Type.Record} type.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class RecordConstructorValue implements Value, CreatesDynamicTypesValue {
+    private static final ObjectPlanHash BASE_HASH = new ObjectPlanHash("Record-Constructor-Value");
+    @Nonnull
+    protected final List<Pair<? extends Value, Optional<String>>> childrenAndNames;
+    @Nonnull
+    private final Supplier<List<? extends Value>> childrenSupplier;
+    @Nonnull
+    private final Supplier<Type.Record> resultTypeSupplier;
+
+    private RecordConstructorValue(@Nonnull List<Pair<? extends Value, Optional<String>>> childrenAndNames) {
+        this.childrenAndNames = ImmutableList.copyOf(childrenAndNames);
+        this.childrenSupplier = Suppliers.memoize(this::computeChildren);
+        this.resultTypeSupplier = Suppliers.memoize(this::computeResultType);
+    }
+
+    @Nonnull
+    @Override
+    public Iterable<? extends Value> getChildren() {
+        return childrenSupplier.get();
+    }
+
+    private List<? extends Value> computeChildren() {
+        return childrenAndNames
+                .stream()
+                .map(Pair::getKey)
+                .collect(ImmutableList.toImmutableList());
+    }
+
+    @Nonnull
+    @Override
+    public Type.Record getResultType() {
+        return resultTypeSupplier.get();
+    }
+
+    @Nonnull
+    public Type.Record computeResultType() {
+        final ImmutableList<Type.Record.Field> fields = childrenAndNames
+                .stream()
+                .map(childAndName -> Type.Record.Field.of(childAndName.getKey().getResultType(), childAndName.getValue()))
+                .collect(ImmutableList.toImmutableList());
+        return Type.Record.fromFields(fields);
+    }
+
+    @Nullable
+    @Override
+    public <M extends Message> Object eval(@Nonnull final FDBRecordStoreBase<M> store, @Nonnull final EvaluationContext context, @Nullable final FDBRecord<M> record, @Nullable final M message) {
+        final DynamicMessage.Builder resultMessageBuilder = newMessageBuilderForType(context.getDynamicSchema());
+        final Descriptors.Descriptor descriptorForType = resultMessageBuilder.getDescriptorForType();
+
+        int i = 0;
+        final List<Type.Record.Field> fields = Objects.requireNonNull(getResultType().getFields());
+
+        for (final Value child : getChildren()) {
+            final Object childResultElement = child.eval(store, context, record, message);
+            if (childResultElement != null) {
+                resultMessageBuilder.setField(descriptorForType.findFieldByNumber(fields.get(i).getFieldIndex()), childResultElement);
+            }
+            i ++;
+        }
+
+        return resultMessageBuilder.build();
+    }
+
+    @Nonnull
+    private DynamicMessage.Builder newMessageBuilderForType(@Nonnull DynamicSchema dynamicSchema) {
+        return dynamicSchema.newMessageBuilder(dynamicSchema.getProtoTypeName(getResultType()));
```

Comment:
Again, I think `newMessageBuilder(Type)` is cleaner.

Reply:
done.